spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/vault/SpringVaultClientConfiguration.java

```diff
@@ -0,0 +1,582 @@
+
+	private final RestOperations externalRestOperations;
+
+	private final Log log = LogFactory.getLog(getClass());
+
+	public SpringVaultClientConfiguration(VaultEnvironmentProperties vaultProperties,
+			ConfigTokenProvider configTokenProvider) {
+		this.vaultProperties = vaultProperties;
+		this.configTokenProvider = configTokenProvider;
+		this.externalRestOperations = new RestTemplate(
+				clientHttpRequestFactoryWrapper().getClientHttpRequestFactory());
+	}
+
+	@Override
+	public VaultEndpoint vaultEndpoint() {
+
+		URI baseUrl = UriComponentsBuilder.newInstance()
+				.scheme(vaultProperties.getScheme()).host(vaultProperties.getHost())
+				.port(vaultProperties.getPort()).build().toUri();
+
+		return VaultEndpoint.from(baseUrl);
+	}
+
+	@Override
+	protected RestTemplateBuilder restTemplateBuilder(
+			VaultEndpointProvider endpointProvider,
+			ClientHttpRequestFactory requestFactory) {
+
+		RestTemplateBuilder restTemplateBuilder = super.restTemplateBuilder(
+				endpointProvider, requestFactory);
+
+		if (vaultProperties.getNamespace() != null) {
+			restTemplateBuilder.customizers(
+					restTemplate -> restTemplate.getInterceptors().add(VaultClients
+							.createNamespaceInterceptor(vaultProperties.getNamespace())));
+		}
+
+		return restTemplateBuilder;
+	}
+
+	@Override
+	public SslConfiguration sslConfiguration() {
+		if (vaultProperties.isSkipSslValidation()) {
+			log.warn("The '" + VAULT_PROPERTIES_PREFIX + "skipSslValidation' property "
+					+ "is not supported by this Vault environment repository implementation. "
+					+ "Use the '" + VAULT_PROPERTIES_PREFIX
+					+ "ssl` properties to provide "
+					+ "custom keyStore and trustStore material instead.");
+		}
+
+		VaultEnvironmentProperties.Ssl ssl = vaultProperties.getSsl();
+
+		SslConfiguration.KeyStoreConfiguration keyStoreConfiguration = getKeyStoreConfiguration(
+				ssl.getKeyStore(), ssl.getKeyStorePassword());
+
+		SslConfiguration.KeyStoreConfiguration trustStoreConfiguration = getKeyStoreConfiguration(
+				ssl.getTrustStore(), ssl.getTrustStorePassword());
+
+		return new SslConfiguration(keyStoreConfiguration, trustStoreConfiguration);
+	}
+
+	private SslConfiguration.KeyStoreConfiguration getKeyStoreConfiguration(
+			Resource resourceProperty, String passwordProperty) {
+
+		if (resourceProperty == null) {
+			return SslConfiguration.KeyStoreConfiguration.unconfigured();
+		}
+
+		if (StringUtils.hasText(passwordProperty)) {
+			return SslConfiguration.KeyStoreConfiguration.of(resourceProperty,
+					passwordProperty.toCharArray());
+		}
+
+		return SslConfiguration.KeyStoreConfiguration.of(resourceProperty);
+	}
+
+	/**
+	 * @return a new {@link ClientAuthentication}.
+	 */
+	public ClientAuthentication clientAuthentication() {
+
+		AuthenticationMethod authentication = this.vaultProperties.getAuthentication();
+
+		if (authentication == null) {
+			return new ConfigTokenProviderAuthentication(configTokenProvider);
+		}
+
+		switch (authentication) {
```

Comment:
I'm a little concerned about the size of this class. I'm ok refactoring it later since it's configuration and not a public api. I'd wonder if something like a `VaultAuthenticationFactory` interface with a `supports(AuthenticationMethod)` and `getClientAuthentication(VaultProperties, RestOperations)` would be helpful. That way each method could be a separate small class, you could also do things like `@ConditionalOnClass`

Reply:
+1 on that. I was trying to stick pretty close to the two classes from Spring Vault and Spring Cloud Vault that do similar configuration to have the best chance of getting something that works in short order, but decomposing this into a set of smaller classes would be better.